In [1]:
from d3rlpy.datasets import get_cartpole

In [20]:
dataset, env = get_cartpole()

In [138]:
dataset.size()

1582

In [89]:
len(dataset.rewards)

99866

In [90]:
len(dataset.episodes)

1582

In [91]:
len(dataset.actions)

99866

In [139]:
from sklearn.model_selection import train_test_split

train_episodes, test_episodes = train_test_split(dataset, test_size=0.2)

In [140]:
from d3rlpy.algos import DQN

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset)

In [141]:
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

from d3rlpy.metrics.scorer import evaluate_on_environment

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

In [142]:
env

<TimeLimit<CartPoleEnv<CartPole-v0>>>

In [143]:
# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:01:41 [debug    ] RoundIterator is selected.
2023-03-13 19:01:41 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190141
2023-03-13 19:01:41 [warning  ] Skip building models since they're already built.
2023-03-13 19:01:41 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190141/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:01:51 [info     ] DQN_20230313190141: epoch=1 step=2500 epoch=1 metrics={'time_sample_batch': 0.00012184162139892578, 'time_algorithm_update': 0.003392033576965332, 'loss': 0.012645094241556217, 'time_step': 0.00361170015335083, 'td_error': 0.9881827529171064, 'value_scale': 1.0015601515105368, 'environment': 9.4} step=2500
2023-03-13 19:01:51 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_2500.pt


Epoch 2/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:00 [info     ] DQN_20230313190141: epoch=2 step=5000 epoch=2 metrics={'time_sample_batch': 0.00010874042510986328, 'time_algorithm_update': 0.0030711585998535154, 'loss': 3.558228413348843e-05, 'time_step': 0.003266118335723877, 'td_error': 0.9870644139670596, 'value_scale': 1.0012083776692806, 'environment': 9.3} step=5000
2023-03-13 19:02:00 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_5000.pt


Epoch 3/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:10 [info     ] DQN_20230313190141: epoch=3 step=7500 epoch=3 metrics={'time_sample_batch': 0.00011303062438964843, 'time_algorithm_update': 0.0031888752937316894, 'loss': 3.1675507657382695e-05, 'time_step': 0.0033857173919677732, 'td_error': 0.9833814717893776, 'value_scale': 0.9939531941963918, 'environment': 17.3} step=7500
2023-03-13 19:02:10 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_7500.pt


Epoch 4/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:20 [info     ] DQN_20230313190141: epoch=4 step=10000 epoch=4 metrics={'time_sample_batch': 0.00011527967453002929, 'time_algorithm_update': 0.0031881381034851073, 'loss': 0.007066348025825027, 'time_step': 0.00339133882522583, 'td_error': 0.9918641160631149, 'value_scale': 2.0060413379784223, 'environment': 12.4} step=10000
2023-03-13 19:02:20 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_10000.pt


Epoch 5/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:29 [info     ] DQN_20230313190141: epoch=5 step=12500 epoch=5 metrics={'time_sample_batch': 0.00010668191909790038, 'time_algorithm_update': 0.003076361846923828, 'loss': 0.006230760553551954, 'time_step': 0.003267509746551514, 'td_error': 0.9752552312302291, 'value_scale': 1.98381710454265, 'environment': 19.9} step=12500
2023-03-13 19:02:29 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_12500.pt


Epoch 6/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:38 [info     ] DQN_20230313190141: epoch=6 step=15000 epoch=6 metrics={'time_sample_batch': 0.00010726566314697266, 'time_algorithm_update': 0.0031132880210876465, 'loss': 0.006048588067226229, 'time_step': 0.00330434045791626, 'td_error': 0.9825717391032092, 'value_scale': 1.9832849040915217, 'environment': 17.8} step=15000
2023-03-13 19:02:38 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_15000.pt


Epoch 7/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:48 [info     ] DQN_20230313190141: epoch=7 step=17500 epoch=7 metrics={'time_sample_batch': 0.00010572671890258789, 'time_algorithm_update': 0.003082563495635986, 'loss': 0.015362253805273212, 'time_step': 0.0032682649612426756, 'td_error': 1.0135379743300548, 'value_scale': 2.99799875456408, 'environment': 175.3} step=17500
2023-03-13 19:02:48 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_17500.pt


Epoch 8/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:02:57 [info     ] DQN_20230313190141: epoch=8 step=20000 epoch=8 metrics={'time_sample_batch': 0.00010021581649780273, 'time_algorithm_update': 0.0030054548263549804, 'loss': 0.01864413274436083, 'time_step': 0.0031805929183959963, 'td_error': 0.9966686411730681, 'value_scale': 2.9763200114003054, 'environment': 114.7} step=20000
2023-03-13 19:02:57 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_20000.pt


Epoch 9/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:03:07 [info     ] DQN_20230313190141: epoch=9 step=22500 epoch=9 metrics={'time_sample_batch': 0.00010303192138671875, 'time_algorithm_update': 0.0030933475494384767, 'loss': 0.01833996931213187, 'time_step': 0.003274539566040039, 'td_error': 0.9954841789499165, 'value_scale': 2.973896644438197, 'environment': 153.2} step=22500
2023-03-13 19:03:07 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_22500.pt


Epoch 10/10:   0%|          | 0/2500 [00:00<?, ?it/s]

2023-03-13 19:03:17 [info     ] DQN_20230313190141: epoch=10 step=25000 epoch=10 metrics={'time_sample_batch': 0.00010579662322998047, 'time_algorithm_update': 0.0032065747261047365, 'loss': 0.024241683905909302, 'time_step': 0.0033939443588256834, 'td_error': 1.0467469998968637, 'value_scale': 3.9858706374229054, 'environment': 199.9} step=25000
2023-03-13 19:03:17 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190141/model_25000.pt


[(1,
  {'time_sample_batch': 0.00012184162139892578,
   'time_algorithm_update': 0.003392033576965332,
   'loss': 0.012645094241556217,
   'time_step': 0.00361170015335083,
   'td_error': 0.9881827529171064,
   'value_scale': 1.0015601515105368,
   'environment': 9.4}),
 (2,
  {'time_sample_batch': 0.00010874042510986328,
   'time_algorithm_update': 0.0030711585998535154,
   'loss': 3.558228413348843e-05,
   'time_step': 0.003266118335723877,
   'td_error': 0.9870644139670596,
   'value_scale': 1.0012083776692806,
   'environment': 9.3}),
 (3,
  {'time_sample_batch': 0.00011303062438964843,
   'time_algorithm_update': 0.0031888752937316894,
   'loss': 3.1675507657382695e-05,
   'time_step': 0.0033857173919677732,
   'td_error': 0.9833814717893776,
   'value_scale': 0.9939531941963918,
   'environment': 17.3}),
 (4,
  {'time_sample_batch': 0.00011527967453002929,
   'time_algorithm_update': 0.0031881381034851073,
   'loss': 0.007066348025825027,
   'time_step': 0.00339133882522583,
   '

### Metrics

**td_error** - larger td error means that the estimated state values are too far from the correct state values
The TD error is the difference between the correct reward based value and our estimate of the reward based value. The policy is the actor and the value function is the critic, with the ‘criticism’ being the TD error. If the TD error is positive the value of the action was greater than expected, suggesting the chosen action should be taken more often. If the TD error was negative the action had a lower value than expected, and so will be done less often in future states which are similar.

**discounted sum of average** - This is the discounted sum of advantage of the _**selected actions**_. This should be small to indicate that the policy selects actions with larger estimated values. This is the difference between the Q values of the dataset actions or the actions selected in each state and the max Q value according to the policy.  https://github.com/takuseno/d3rlpy/issues/99



In [51]:
observation = env.reset()

print(observation)

# return actions based on the greedy-policy
action = dqn.predict([observation])[0]

# estimate action-values
value = dqn.predict_value([observation], [action])

In [21]:
# This is the most useful metric for evaluating the progress of learning
def evaluate_mean_rewards(dqn):
    evaluate_scorer = evaluate_on_environment(env, n_trials=50)
    mean_episode_return = evaluate_scorer(dqn)
    return(mean_episode_return)

In [22]:
evaluate_mean_rewards(dqn)

NameError: name 'dqn' is not defined

In [54]:
value

In [49]:
dqn.predict([observation])

array([0])

### Generate offline data

In [145]:
import d3rlpy

# setup algorithm
random_policy = d3rlpy.algos.DiscreteRandomPolicy()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=500000, env=env)

# start data collection
random_policy.collect(env, buffer, n_steps=500000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("random_policy_dataset_500000.h5")

2023-03-13 19:04:24 [debug    ] Building model...
2023-03-13 19:04:24 [debug    ] Model has been built.


  0%|          | 0/500000 [00:00<?, ?it/s]

In [33]:
ls

Intro_d3rlpy.ipynb                 random_policy_corridor_dataset.h5
d3rlpy_data/                       random_policy_dataset.h5
d3rlpy_logs/                       random_policy_dataset_80000.h5
poetry.lock                        trained_policy_dataset.h5
pyproject.toml                     trained_policy_dataset_80000.h5


In [146]:
# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=30000, env=env)

# start data collection
dqn.collect(env, buffer, n_steps=30000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("trained_policy_dataset_30000.h5")

2023-03-13 19:04:58 [warning  ] Skip building models since they're already built.


  0%|          | 0/30000 [00:00<?, ?it/s]

import numpy as np

buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=80000, env=env)
dataset2 = buffer.to_mdp_dataset()
dataset2.load("random_policy_dataset_80000.h5")

buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=80000, env=env)
dataset3 = buffer.to_mdp_dataset()
dataset3.load("trained_policy_dataset_80000.h5")

In [147]:
dataset_trained = dataset.load("trained_policy_dataset_30000.h5")
dataset_random = dataset.load("random_policy_dataset_500000.h5")

In [148]:
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.2)

In [149]:
dataset_trained.compute_stats()

{'return': {'mean': 197.7616,
  'std': 12.035108,
  'min': 51.0,
  'max': 199.0,
  'histogram': (array([  1,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
            0,   0,   0,   0,   0,   1, 149]),
   array([ 51. ,  58.4,  65.8,  73.2,  80.6,  88. ,  95.4, 102.8, 110.2,
          117.6, 125. , 132.4, 139.8, 147.2, 154.6, 162. , 169.4, 176.8,
          184.2, 191.6, 199. ], dtype=float32))},
 'reward': {'mean': 1.0,
  'std': 0.0,
  'min': 1.0,
  'max': 1.0,
  'histogram': (array([    0,     0,     0,     0,     0,     0,     0,     0,     0,
              0, 29862,     0,     0,     0,     0,     0,     0,     0,
              0,     0]),
   array([0.5 , 0.55, 0.6 , 0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 , 0.95, 1.  ,
          1.05, 1.1 , 1.15, 1.2 , 1.25, 1.3 , 1.35, 1.4 , 1.45, 1.5 ],
         dtype=float32))},
 'action': {'histogram': [[15465, 14397], array([0, 1])]},
 'observation': {'mean': array([-0.5493725 , -0.44689432, -0.0343954 , -0.02737486], dtype=float32),
  

In [150]:
dataset_random.compute_stats()

{'return': {'mean': 22.354317,
  'std': 11.985191,
  'min': 8.0,
  'max': 119.0,
  'histogram': (array([4730, 7077, 3742, 2736, 1372, 1043,  566,  433,  233,  155,  106,
            61,   38,   34,   17,   10,    4,    3,    4,    3]),
   array([  8.  ,  13.55,  19.1 ,  24.65,  30.2 ,  35.75,  41.3 ,  46.85,
           52.4 ,  57.95,  63.5 ,  69.05,  74.6 ,  80.15,  85.7 ,  91.25,
           96.8 , 102.35, 107.9 , 113.45, 119.  ], dtype=float32))},
 'reward': {'mean': 1.0,
  'std': 0.0,
  'min': 1.0,
  'max': 1.0,
  'histogram': (array([     0,      0,      0,      0,      0,      0,      0,      0,
               0,      0, 499999,      0,      0,      0,      0,      0,
               0,      0,      0,      0]),
   array([0.5 , 0.55, 0.6 , 0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 , 0.95, 1.  ,
          1.05, 1.1 , 1.15, 1.2 , 1.25, 1.3 , 1.35, 1.4 , 1.45, 1.5 ],
         dtype=float32))},
 'action': {'histogram': [[249843, 250156], array([0, 1])]},
 'observation': {'mean': array([2.318179

In [101]:
dataset_trained.actions

array([0, 0, 0, ..., 1, 0, 0], dtype=int32)

In [76]:
dataset_random.actions

array([1, 1, 1, ..., 0, 0, 0], dtype=int32)

In [57]:
dataset_trained.rewards

array([1., 1., 1., ..., 1., 1., 1.], dtype=float32)

In [152]:
len(dataset_trained.episodes)

151

In [151]:
len(dataset_random.episodes)

22367

In [79]:
dataset_trained.observations

array([[ 0.02962353, -0.0323091 , -0.03679555,  0.03139311],
       [ 0.02897735, -0.22688459, -0.03616769,  0.31224334],
       [ 0.02443966, -0.42147312, -0.02992282,  0.5933043 ],
       ...,
       [ 0.2347083 ,  0.5337355 , -0.13025276, -0.45787495],
       [ 0.24538301,  0.34067228, -0.13941026, -0.20891936],
       [ 0.25219646,  0.14779048, -0.14358865,  0.03674239]],
      dtype=float32)

### Train on offline data

In [153]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset_trained)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:06:06 [debug    ] RoundIterator is selected.
2023-03-13 19:06:06 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190606
2023-03-13 19:06:06 [warning  ] Skip building models since they're already built.
2023-03-13 19:06:06 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190606/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:09 [info     ] DQN_20230313190606: epoch=1 step=742 epoch=1 metrics={'time_sample_batch': 0.00010505551597821423, 'time_algorithm_update': 0.002879885329069153, 'loss': 0.031179017385418388, 'time_step': 0.003060433421173507, 'td_error': 0.9846738294130843, 'advantage': -0.1300933121034453, 'value_scale': 1.0728607263069005, 'environment': 86.4} step=742
2023-03-13 19:06:09 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_742.pt


Epoch 2/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:12 [info     ] DQN_20230313190606: epoch=2 step=1484 epoch=2 metrics={'time_sample_batch': 0.00011077659792013246, 'time_algorithm_update': 0.0030716243137246515, 'loss': 3.344963113403049e-05, 'time_step': 0.003261560699689099, 'td_error': 0.985295856526353, 'advantage': -0.1917720939016169, 'value_scale': 1.0729103422340371, 'environment': 120.2} step=1484
2023-03-13 19:06:12 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_1484.pt


Epoch 3/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:15 [info     ] DQN_20230313190606: epoch=3 step=2226 epoch=3 metrics={'time_sample_batch': 0.00010746990574016726, 'time_algorithm_update': 0.003062619353240391, 'loss': 1.5665062441480843e-05, 'time_step': 0.0032469680045814206, 'td_error': 0.9840360184373476, 'advantage': -0.15674497508117402, 'value_scale': 1.0731224996386852, 'environment': 123.9} step=2226
2023-03-13 19:06:15 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_2226.pt


Epoch 4/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:18 [info     ] DQN_20230313190606: epoch=4 step=2968 epoch=4 metrics={'time_sample_batch': 0.00010607602461329047, 'time_algorithm_update': 0.003131949676657623, 'loss': 1.067789256195109e-05, 'time_step': 0.0033144237538874953, 'td_error': 0.9840877200697233, 'advantage': -0.15886232738107622, 'value_scale': 1.073022628412695, 'environment': 65.3} step=2968
2023-03-13 19:06:18 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_2968.pt


Epoch 5/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:21 [info     ] DQN_20230313190606: epoch=5 step=3710 epoch=5 metrics={'time_sample_batch': 0.00010663704730751059, 'time_algorithm_update': 0.0030635277216003913, 'loss': 8.527066199344161e-06, 'time_step': 0.0032455982223675256, 'td_error': 0.9836855233927463, 'advantage': -0.14836029094288894, 'value_scale': 1.073164146328116, 'environment': 81.3} step=3710
2023-03-13 19:06:21 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_3710.pt


Epoch 6/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:24 [info     ] DQN_20230313190606: epoch=6 step=4452 epoch=6 metrics={'time_sample_batch': 0.00010828509163663715, 'time_algorithm_update': 0.0030746199692677297, 'loss': 7.0954519173631595e-06, 'time_step': 0.003260059176750903, 'td_error': 0.9831215379138788, 'advantage': -0.12544745650668326, 'value_scale': 1.0730089680529256, 'environment': 99.3} step=4452
2023-03-13 19:06:24 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_4452.pt


Epoch 7/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:27 [info     ] DQN_20230313190606: epoch=7 step=5194 epoch=7 metrics={'time_sample_batch': 0.00010595553004838065, 'time_algorithm_update': 0.0029992861246484308, 'loss': 7.0120120548852535e-06, 'time_step': 0.0031801108722738177, 'td_error': 0.982881306147701, 'advantage': -0.11057738133555271, 'value_scale': 1.0711203961692182, 'environment': 152.6} step=5194
2023-03-13 19:06:27 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_5194.pt


Epoch 8/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:30 [info     ] DQN_20230313190606: epoch=8 step=5936 epoch=8 metrics={'time_sample_batch': 0.00011502250506871473, 'time_algorithm_update': 0.003311586187213579, 'loss': 7.016652208064585e-06, 'time_step': 0.003509262180071314, 'td_error': 0.9860360994536276, 'advantage': -0.22346773301398976, 'value_scale': 1.0732343780454097, 'environment': 149.1} step=5936
2023-03-13 19:06:30 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_5936.pt


Epoch 9/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:34 [info     ] DQN_20230313190606: epoch=9 step=6678 epoch=9 metrics={'time_sample_batch': 0.0001122112865396587, 'time_algorithm_update': 0.0031580108837939978, 'loss': 7.5880011481197564e-06, 'time_step': 0.003349788748029107, 'td_error': 0.9845790799986306, 'advantage': -0.17504556564075838, 'value_scale': 1.0736828844250994, 'environment': 169.0} step=6678
2023-03-13 19:06:34 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_6678.pt


Epoch 10/10:   0%|          | 0/742 [00:00<?, ?it/s]

2023-03-13 19:06:37 [info     ] DQN_20230313190606: epoch=10 step=7420 epoch=10 metrics={'time_sample_batch': 0.00011082286783305782, 'time_algorithm_update': 0.0030986314835252785, 'loss': 7.565187162005304e-06, 'time_step': 0.0032873475005363198, 'td_error': 0.9839800890170002, 'advantage': -0.1490605409202769, 'value_scale': 1.0740309759892235, 'environment': 189.7} step=7420
2023-03-13 19:06:37 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190606/model_7420.pt


[(1,
  {'time_sample_batch': 0.00010505551597821423,
   'time_algorithm_update': 0.002879885329069153,
   'loss': 0.031179017385418388,
   'time_step': 0.003060433421173507,
   'td_error': 0.9846738294130843,
   'advantage': -0.1300933121034453,
   'value_scale': 1.0728607263069005,
   'environment': 86.4}),
 (2,
  {'time_sample_batch': 0.00011077659792013246,
   'time_algorithm_update': 0.0030716243137246515,
   'loss': 3.344963113403049e-05,
   'time_step': 0.003261560699689099,
   'td_error': 0.985295856526353,
   'advantage': -0.1917720939016169,
   'value_scale': 1.0729103422340371,
   'environment': 120.2}),
 (3,
  {'time_sample_batch': 0.00010746990574016726,
   'time_algorithm_update': 0.003062619353240391,
   'loss': 1.5665062441480843e-05,
   'time_step': 0.0032469680045814206,
   'td_error': 0.9840360184373476,
   'advantage': -0.15674497508117402,
   'value_scale': 1.0731224996386852,
   'environment': 123.9}),
 (4,
  {'time_sample_batch': 0.00010607602461329047,
   'time_a

In [154]:
evaluate_mean_rewards(dqn)

118.18

In [155]:
train_episodes, test_episodes = train_test_split(dataset_random, test_size=0.2)

In [156]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import evaluate_on_environment

# if you don't use GPU, set use_gpu=False instead.
dqn = DQN(use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn.build_with_dataset(dataset_random)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-13 19:07:50 [debug    ] RoundIterator is selected.
2023-03-13 19:07:50 [info     ] Directory is created at d3rlpy_logs/DQN_20230313190750
2023-03-13 19:07:50 [warning  ] Skip building models since they're already built.
2023-03-13 19:07:50 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230313190750/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (4,), 'action_size': 2}


Epoch 1/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:08:41 [info     ] DQN_20230313190750: epoch=1 step=12473 epoch=1 metrics={'time_sample_batch': 0.00010740599291637906, 'time_algorithm_update': 0.0029705155300059948, 'loss': 0.007300398724316957, 'time_step': 0.003226995257877132, 'td_error': 0.9154510430734613, 'value_scale': 1.989812987434266, 'environment': 13.1} step=12473
2023-03-13 19:08:41 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_12473.pt


Epoch 2/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:09:30 [info     ] DQN_20230313190750: epoch=2 step=24946 epoch=2 metrics={'time_sample_batch': 9.262298969949896e-05, 'time_algorithm_update': 0.002876004726972957, 'loss': 0.019224161197751933, 'time_step': 0.0030891139496015984, 'td_error': 0.8774920830094897, 'value_scale': 3.8550574445173433, 'environment': 165.1} step=24946
2023-03-13 19:09:30 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_24946.pt


Epoch 3/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:10:22 [info     ] DQN_20230313190750: epoch=3 step=37419 epoch=3 metrics={'time_sample_batch': 9.637766657400884e-05, 'time_algorithm_update': 0.002961667989141947, 'loss': 0.025945931547490222, 'time_step': 0.0031868134792734336, 'td_error': 0.8805363140884244, 'value_scale': 4.753485571590386, 'environment': 129.4} step=37419
2023-03-13 19:10:22 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_37419.pt


Epoch 4/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:11:14 [info     ] DQN_20230313190750: epoch=4 step=49892 epoch=4 metrics={'time_sample_batch': 0.00010250079778033443, 'time_algorithm_update': 0.002999999839894651, 'loss': 0.029049355781752006, 'time_step': 0.0032261638990010904, 'td_error': 0.8977108749277593, 'value_scale': 6.4617853311665225, 'environment': 198.4} step=49892
2023-03-13 19:11:14 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_49892.pt


Epoch 5/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:12:07 [info     ] DQN_20230313190750: epoch=5 step=62365 epoch=5 metrics={'time_sample_batch': 0.00010686229227788273, 'time_algorithm_update': 0.0030226505797627855, 'loss': 0.033463577602898954, 'time_step': 0.003262710104888571, 'td_error': 0.9232798758382001, 'value_scale': 7.402156132463645, 'environment': 189.9} step=62365
2023-03-13 19:12:07 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_62365.pt


Epoch 6/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:13:02 [info     ] DQN_20230313190750: epoch=6 step=74838 epoch=6 metrics={'time_sample_batch': 0.00010990138846392583, 'time_algorithm_update': 0.0032613820294891166, 'loss': 0.04279645937602827, 'time_step': 0.003503515450275672, 'td_error': 1.022839158552869, 'value_scale': 9.176658419550185, 'environment': 148.8} step=74838
2023-03-13 19:13:02 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_74838.pt


Epoch 7/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:13:55 [info     ] DQN_20230313190750: epoch=7 step=87311 epoch=7 metrics={'time_sample_batch': 0.00010684241291267262, 'time_algorithm_update': 0.0031818968726171674, 'loss': 0.05209862822889459, 'time_step': 0.0034250694489905325, 'td_error': 1.0755130054085789, 'value_scale': 9.98658522042752, 'environment': 115.0} step=87311
2023-03-13 19:13:55 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_87311.pt


Epoch 8/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:14:47 [info     ] DQN_20230313190750: epoch=8 step=99784 epoch=8 metrics={'time_sample_batch': 0.00010092547277607915, 'time_algorithm_update': 0.0030981286109368305, 'loss': 0.06506972424293393, 'time_step': 0.0033279058211293034, 'td_error': 1.195821588186276, 'value_scale': 11.665359968979724, 'environment': 122.1} step=99784
2023-03-13 19:14:47 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_99784.pt


Epoch 9/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:15:40 [info     ] DQN_20230313190750: epoch=9 step=112257 epoch=9 metrics={'time_sample_batch': 0.00010414751646637793, 'time_algorithm_update': 0.0031380276820295117, 'loss': 0.07419467414129627, 'time_step': 0.0033796215691983553, 'td_error': 1.3563738346002665, 'value_scale': 13.410262844828551, 'environment': 112.3} step=112257
2023-03-13 19:15:40 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_112257.pt


Epoch 10/10:   0%|          | 0/12473 [00:00<?, ?it/s]

2023-03-13 19:16:33 [info     ] DQN_20230313190750: epoch=10 step=124730 epoch=10 metrics={'time_sample_batch': 0.0001025821884890504, 'time_algorithm_update': 0.003151704972015677, 'loss': 0.08543142601732101, 'time_step': 0.0033856501969608357, 'td_error': 1.3679511803102369, 'value_scale': 14.214068009820659, 'environment': 113.1} step=124730
2023-03-13 19:16:33 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230313190750/model_124730.pt


[(1,
  {'time_sample_batch': 0.00010740599291637906,
   'time_algorithm_update': 0.0029705155300059948,
   'loss': 0.007300398724316957,
   'time_step': 0.003226995257877132,
   'td_error': 0.9154510430734613,
   'value_scale': 1.989812987434266,
   'environment': 13.1}),
 (2,
  {'time_sample_batch': 9.262298969949896e-05,
   'time_algorithm_update': 0.002876004726972957,
   'loss': 0.019224161197751933,
   'time_step': 0.0030891139496015984,
   'td_error': 0.8774920830094897,
   'value_scale': 3.8550574445173433,
   'environment': 165.1}),
 (3,
  {'time_sample_batch': 9.637766657400884e-05,
   'time_algorithm_update': 0.002961667989141947,
   'loss': 0.025945931547490222,
   'time_step': 0.0031868134792734336,
   'td_error': 0.8805363140884244,
   'value_scale': 4.753485571590386,
   'environment': 129.4}),
 (4,
  {'time_sample_batch': 0.00010250079778033443,
   'time_algorithm_update': 0.002999999839894651,
   'loss': 0.029049355781752006,
   'time_step': 0.0032261638990010904,
   't

In [157]:
evaluate_mean_rewards(dqn)

112.54

### Online learning

In [173]:
# This setup works and is used for online collection of data
from gym.wrappers import TimeLimit
from d3rlpy.algos import DQN
from d3rlpy.online.buffers import ReplayBuffer
from d3rlpy.online.explorers import LinearDecayEpsilonGreedy, ConstantEpsilonGreedy

config={"corridor_length": 5}
env_corridor = SimpleCorridor_d3rlpy(config=config)

env = TimeLimit(env_corridor, max_episode_steps=2000)

#env = gym.make('CartPole-v0')
#eval_env = gym.make('CartPole-v0')

# setup algorithm
dqn = DQN(batch_size=128,
          learning_rate=2.5e-5,
          target_update_interval=100,
          use_gpu=False)

# setup replay buffer
buffer = ReplayBuffer(maxlen=100000, env=env)

# setup explorers
explorer = LinearDecayEpsilonGreedy(start_epsilon=1.0,
                                    end_epsilon=0.1,
                                    duration=10000)

# start training
dqn.fit_online(env,
               buffer,
               explorer=explorer, # you don't need this with probablistic policy algorithms
               eval_env=env,
               n_steps=6000, # the number of total steps to train.
               n_steps_per_epoch=1000,
               update_interval=10) # update parameters every 10 steps.

# export online dataset as MDPDataset
dataset_online = buffer.to_mdp_dataset()

# save MDPDataset
dataset_online.dump("trained_corridor_policy_dataset_10000_online.h5")

2023-03-14 18:24:19 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230314182419
2023-03-14 18:24:19 [debug    ] Building model...
2023-03-14 18:24:19 [debug    ] Model has been built.
2023-03-14 18:24:19 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230314182419/params.json params={'action_scaler': None, 'batch_size': 128, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 2.5e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/6000 [00:00<?, ?it/s]

2023-03-14 18:24:20 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314182419/model_1000.pt
2023-03-14 18:24:20 [info     ] DQN_online_20230314182419: epoch=1 step=1000 epoch=1 metrics={'time_inference': 0.0004101440906524658, 'time_environment_step': 8.50057601928711e-06, 'time_step': 0.0008164141178131104, 'rollout_return': 1.4146739130434787, 'time_sample_batch': 0.0001660477031360973, 'time_algorithm_update': 0.003912180662155151, 'loss': 0.12462697334756906, 'evaluation': 1.6} step=1000
2023-03-14 18:24:21 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314182419/model_2000.pt
2023-03-14 18:24:21 [info     ] DQN_online_20230314182419: epoch=2 step=2000 epoch=2 metrics={'time_inference': 0.00040985918045043947, 'time_environment_step': 8.703947067260742e-06, 'rollout_return': 1.707203389830509, 'time_step': 0.0008667199611663818, 'time_sample_batch': 0.00017389297485351563, 'time_algorithm_update': 0.0038977885246276856, 'loss': 0.0903030

In [172]:
# This setup works and is used for online collection of data
# setup replay buffer
buffer = ReplayBuffer(maxlen=100000, env=env)

# setup explorers
explorer = LinearDecayEpsilonGreedy(start_epsilon=1.0,
                                    end_epsilon=0.1,
                                    duration=10000)

# start training
dqn.fit_online(env,
               buffer,
               explorer=explorer, # you don't need this with probablistic policy algorithms
               eval_env=env,
               n_steps=6000, # the number of total steps to train.
               n_steps_per_epoch=1000,
               update_interval=10) # update parameters every 10 steps.

# export online dataset as MDPDataset
dataset_online = buffer.to_mdp_dataset()

# save MDPDataset
dataset_online.dump("trained_corridor_policy_dataset_10000_online.h5")

2023-03-14 18:23:41 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230314182341
2023-03-14 18:23:41 [warning  ] Skip building models since they're already built.
2023-03-14 18:23:41 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230314182341/params.json params={'action_scaler': None, 'batch_size': 128, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 2.5e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/6000 [00:00<?, ?it/s]

2023-03-14 18:23:42 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314182341/model_1000.pt
2023-03-14 18:23:42 [info     ] DQN_online_20230314182341: epoch=1 step=1000 epoch=1 metrics={'time_inference': 0.0004189438819885254, 'time_environment_step': 8.656978607177735e-06, 'time_step': 0.0008457856178283691, 'rollout_return': 1.354255319148936, 'time_sample_batch': 0.00016421621496027166, 'time_algorithm_update': 0.004126462069424716, 'loss': 0.13059583873572675, 'evaluation': 2.3999999999999995} step=1000
2023-03-14 18:23:43 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314182341/model_2000.pt
2023-03-14 18:23:43 [info     ] DQN_online_20230314182341: epoch=2 step=2000 epoch=2 metrics={'time_inference': 0.00042982983589172363, 'time_environment_step': 8.999347686767579e-06, 'rollout_return': 1.7503906250000005, 'time_step': 0.0009194846153259277, 'time_sample_batch': 0.00017924785614013673, 'time_algorithm_update': 0.0041862988471984865, 

In [185]:
dqn.save_model('Online_trained_corridor_6000.pt')

In [187]:
dqn_loaded = DQN()
dqn_loaded.build_with_dataset(dataset_online)
dqn_loaded.load_model('Online_trained_corridor_6000.pt')

In [175]:
for episode in dataset_online.episodes:
    print(episode.actions, episode.compute_return())

[4 2 4 4] 2.35
[3 1 0 4 3 1 2 2 3 1 0 0 3 0 1 2 0 4 1 1] 0.24999988
[0 2 2 2 2 1 1 0 0 0 0 2 4 1 1 3 4] 1.15
[1 3 1 0 0 1 1 1 1 0 0 0 2 3 4 4] 0.35000002
[0 4 1 3 4] 1.6
[3 1 4 2 0 1 4] 0.79999995
[0 3 4 4 2 4] 2.25
[0 4 3 4 2 2 4] 2.2
[2 3 4 4 1] 2.3
[2 1 0 1 1 0 3 1 2 1 0 0 3 2 1 0 2 0 2 3 1 0 3 3 4 1 3 4] -0.14999986
[1 0 2 3 2 2 2 4 2 4 4] 2.0
[0 4 2 1 3 3 3 3 2 2 4] 1.3
[1 0 4 2 4 2 2 1] 2.15
[2 3 0 3 2 4 2 2 4 0 0 4 4] 1.85
[3 0 3 0 2 0 3 1 1 4 3 3 2 2 1] 1.85
[3 4 3 2 3 0 2 2 4 0 3 0 3 3 1 3 1 1 3 0 4] 0.05000007
[4 1 2 2 3 4] 1.55
[1 2 0 4 0 3 2 1 3 4 0 0 3 4 4] 1.75
[4 3 4 0 3 3 4] 2.05
[1 0 4 4 2 2 4] 2.2
[4 1 1 0 0 3 3 1 4] 0.79999995
[3 3 4 0 1 1 3 2 4] 1.4
[1 1 0 3 0 3 1 0 4 1 3 0 0 0 2 2 0 1 2 0 1 0 0 3 4 0 0 0 3 0 0 1 0 1 3 1 4
 1] -2.3841858e-07
[4 1 4] 1.7
[2 4 1 4] 1.65
[3 2 3 2 3 1 4 2 1 0 4] 1.2
[2 1 1 1 3 1 4] 1.7
[4 2 0 1 2 2 2 4 1] 2.1
[1 4 4] 1.6
[4 3 0 1 3 1 1 4] 1.5999999
[3 4 2 1 2 1 2 1] 1.6
[0 4 3 3 4 3 0 2 4] 1.95
[2 1 1 0 0 2 0 2 2 1 2 2 2 4 2 3 2 3 2 3 0

In [189]:
from d3rlpy.metrics.scorer import evaluate_on_environment
evaluate_scorer = evaluate_on_environment(env, n_trials=50)
mean_episode_return = evaluate_scorer(dqn)
mean_episode_return

2.4000000000000004

In [191]:
for i in range(5):
    obs = np.array([i])
    action = dqn.predict([obs])[0]
    print(action)

4
1
4
4
4


### Export data using the trained environment

In [197]:
import d3rlpy

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=30000, env=env)

# start data collection
dqn_loaded.collect(env, buffer, explorer=explorer, deterministic=False, n_steps=10000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("trained_corridor_policy_dataset_10000.h5")

2023-03-14 18:39:42 [warning  ] Skip building models since they're already built.


  0%|          | 0/10000 [00:00<?, ?it/s]

In [198]:
for episode in dataset.episodes:
    print(episode.actions)

[0 0 4 2 2 0 2 0 0 3 0 3 2 0 1 1 3 1 2 0 4 1]
[3 0 2 3 2 1 2 1 0 3 3 0 0 0 3 3 2 4 4 1]
[2 0 0 2 3 0 0 4 4 1]
[4 4 1]
[1 0 0 0 0 2 4 0 2 2 4 3 2 0 0 2 1 0 1 4 4]
[2 4 0 0 0 0 4 2 1 3 4]
[4 3 3 1 3 2 4]
[1 2 0 1 1 3 0 4 2 2 1 3 4]
[4 4 4]
[1 3 0 2 1 0 0 4 2 4 4]
[2 1 0 4 4 3 1]
[3 0 3 1 4 0 4 0 0 1 3 0 3 4 1]
[2 2 3 3 4 1 2 0 4 4]
[0 3 3 0 2 0 0 2 4 1 3 0 1 0 4 1]
[3 2 3 1 3 3 1 0 2 0 0 1 3 2 2 0 0 1 1 2 4 1]
[2 2 2 0 2 0 0 1 1 2 4 4]
[1 4 1 3 0 3 2 2 4]
[0 2 0 1 2 2 1 2 0 0 3 3 1 0 0 4 3 3 3 3 0 4 4]
[2 3 0 0 4 2 3 3 1 4]
[0 1 4 2 1 0 3 2 0 0 4 2 3 1 0 4]
[2 1 1 0 4 1 2 0 1 4]
[3 4 1 1 0 1 2 4]
[3 0 1 4 4]
[0 1 4 3 1 3 3 0 1 3 1]
[1 1 4 2 4]
[4 0 1 0 4 1 2 1]
[0 4 2 2 3 1 3 1 0 2 1 2 0 2 1 3 3 4]
[1 1 3 1 1 2 1]
[2 2 3 3 1 3 2 4 4]
[3 2 3 0 3 4 2 1 3 1 1]
[0 0 2 2 4 0 2 4 3 3 3 4]
[2 0 1 4 4]
[3 1 0 1 4 4]
[1 2 0 1 4 4]
[4 3 1 4]
[1 4 2 3 3 4]
[2 1 3 0 4 1 1 4]
[1 2 3 3 4 2 4]
[3 3 3 2 2 2 1 3 3 2 0 1 1 0 1 0 4 0 2 2 0 3 4 4]
[0 2 3 4 3 3 2 3 0 3 1 3 0 2 1 4 2 3 2 4]
[2 0 4 3 2 3 4 0 2

In [199]:
### Use trained_corridor_policy_dataset_10000_online.h5 which actualy has 6000 samples
### This was generated from a model that predicted the correct actions at each step
#dataset_trained = dataset.load("trained_corridor_policy_dataset_10000_online.h5")
dataset_trained = dataset.load("trained_corridor_policy_dataset_10000.h5")

In [200]:
from sklearn.model_selection import train_test_split

train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.3)

In [201]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer

# if you don't use GPU, set use_gpu=False instead.
dqn1 = DQN(batch_size=32,
          learning_rate=2.5e-4,
          target_update_interval=100,
          use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn1.build_with_dataset(dataset_trained)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn1, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn1)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn1.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=5,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-14 18:40:53 [debug    ] RoundIterator is selected.
2023-03-14 18:40:53 [info     ] Directory is created at d3rlpy_logs/DQN_20230314184053
2023-03-14 18:40:53 [warning  ] Skip building models since they're already built.
2023-03-14 18:40:53 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230314184053/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 0.00025, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


Epoch 1/5:   0%|          | 0/220 [00:00<?, ?it/s]

2023-03-14 18:41:02 [info     ] DQN_20230314184053: epoch=1 step=220 epoch=1 metrics={'time_sample_batch': 0.00011512582952325994, 'time_algorithm_update': 0.0031776298176158556, 'loss': 0.062118962914047934, 'time_step': 0.0033700303597883744, 'td_error': 0.5871389365685559, 'advantage': -2.0706386744446803, 'value_scale': 2.430105906379101, 'environment': -99.49999999999649} step=220
2023-03-14 18:41:02 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314184053/model_220.pt


Epoch 2/5:   0%|          | 0/220 [00:00<?, ?it/s]

2023-03-14 18:41:11 [info     ] DQN_20230314184053: epoch=2 step=440 epoch=2 metrics={'time_sample_batch': 0.00011647722937844017, 'time_algorithm_update': 0.0031397765333002264, 'loss': 0.02544260872633789, 'time_step': 0.003333339907906272, 'td_error': 0.013971651610236935, 'advantage': -0.47265597268204, 'value_scale': 2.4072587970173807, 'environment': -99.49999999999649} step=440
2023-03-14 18:41:11 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314184053/model_440.pt


Epoch 3/5:   0%|          | 0/220 [00:00<?, ?it/s]

2023-03-14 18:41:20 [info     ] DQN_20230314184053: epoch=3 step=660 epoch=3 metrics={'time_sample_batch': 0.00010830489071932706, 'time_algorithm_update': 0.0030313665216619317, 'loss': 0.005230885771478907, 'time_step': 0.003212973204526034, 'td_error': 0.0075467104878888065, 'advantage': -0.1721765102004461, 'value_scale': 2.2782868387756086, 'environment': -99.99999999999646} step=660
2023-03-14 18:41:20 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314184053/model_660.pt


Epoch 4/5:   0%|          | 0/220 [00:00<?, ?it/s]

2023-03-14 18:41:29 [info     ] DQN_20230314184053: epoch=4 step=880 epoch=4 metrics={'time_sample_batch': 0.00010917186737060546, 'time_algorithm_update': 0.0032864364710721104, 'loss': 0.002176166901435036, 'time_step': 0.003469437902623957, 'td_error': 0.0046439071331167, 'advantage': -0.22608734866683156, 'value_scale': 2.214039239053433, 'environment': -99.99999999999646} step=880
2023-03-14 18:41:29 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314184053/model_880.pt


Epoch 5/5:   0%|          | 0/220 [00:00<?, ?it/s]

2023-03-14 18:41:32 [info     ] DQN_20230314184053: epoch=5 step=1100 epoch=5 metrics={'time_sample_batch': 0.00011647614565762607, 'time_algorithm_update': 0.0034429116682572797, 'loss': 0.001956782040956155, 'time_step': 0.003634798526763916, 'td_error': 0.009715166091446133, 'advantage': -0.44852525064988724, 'value_scale': 2.3186216766923766, 'environment': 2.3999999999999995} step=1100
2023-03-14 18:41:32 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314184053/model_1100.pt


[(1,
  {'time_sample_batch': 0.00011512582952325994,
   'time_algorithm_update': 0.0031776298176158556,
   'loss': 0.062118962914047934,
   'time_step': 0.0033700303597883744,
   'td_error': 0.5871389365685559,
   'advantage': -2.0706386744446803,
   'value_scale': 2.430105906379101,
   'environment': -99.49999999999649}),
 (2,
  {'time_sample_batch': 0.00011647722937844017,
   'time_algorithm_update': 0.0031397765333002264,
   'loss': 0.02544260872633789,
   'time_step': 0.003333339907906272,
   'td_error': 0.013971651610236935,
   'advantage': -0.47265597268204,
   'value_scale': 2.4072587970173807,
   'environment': -99.49999999999649}),
 (3,
  {'time_sample_batch': 0.00010830489071932706,
   'time_algorithm_update': 0.0030313665216619317,
   'loss': 0.005230885771478907,
   'time_step': 0.003212973204526034,
   'td_error': 0.0075467104878888065,
   'advantage': -0.1721765102004461,
   'value_scale': 2.2782868387756086,
   'environment': -99.99999999999646}),
 (4,
  {'time_sample_ba

In [202]:
from d3rlpy.metrics.scorer import evaluate_on_environment
evaluate_scorer = evaluate_on_environment(env, n_trials=50)
mean_episode_return = evaluate_scorer(dqn1)
mean_episode_return

2.4000000000000004

In [203]:
for i in range(5):
    obs = np.array([i])
    action = dqn1.predict([obs])[0]
    print(action)

4
1
4
4
1


### Summary

Samples from the random set do not result in good learning. You always want to generate samples using a trained policy. About 30000 samples (151 episodes) from a well trained policy gets better results than 500000 samples (22000 episodes) from a random policy. About 10x more samples and and more than 100x more samples in the random sampling. *More episodes are necessary for random samples because the mean length of these samples are small since they do not take ideal actions in this environment*. This behavior depends on the environment, the corridor example might have longer episodes in the random sampling case. The reward is really the only good metric to measure the performance of a learned algorithm.

Use the dataset summary to estimate the mean reward in both cases. The mean reward in trained is close to 200 while the random scenario is close to 110.

In [205]:
from d3rlpy.dataset import MDPDataset
from d3rlpy.algos import DiscreteCQL
from d3rlpy.datasets import get_cartpole
from sklearn.model_selection import train_test_split
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer

#dataset, env = get_cartpole()

config={"corridor_length": 5}
env_corridor = SimpleCorridor_d3rlpy(config=config)

env = TimeLimit(env_corridor, max_episode_steps=2000)

# setup CQL algorithm (discrete version)
cql = DiscreteCQL(n_frames=4, use_gpu=False)

# split train and test episodes
#dataset_trained = dataset.load("trained_policy_dataset_80000.h5")
dataset_trained = dataset.load("trained_corridor_policy_dataset_10000.h5")
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.2)

# start training
cql.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'environment': evaluate_on_environment(env), # Cartpole environment
            'advantage': discounted_sum_of_advantage_scorer, # smaller is better
            'td_error': td_error_scorer, # smaller is better
            'value_scale': average_value_estimation_scorer # smaller is better
        })

2023-03-14 18:45:15 [debug    ] RoundIterator is selected.
2023-03-14 18:45:15 [info     ] Directory is created at d3rlpy_logs/DiscreteCQL_20230314184515
2023-03-14 18:45:15 [debug    ] Building models...
2023-03-14 18:45:15 [debug    ] Models have been built.
2023-03-14 18:45:15 [info     ] Parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/params.json params={'action_scaler': None, 'alpha': 1.0, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 4, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DiscreteCQL', 'observation

Epoch 1/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:17 [info     ] DiscreteCQL_20230314184515: epoch=1 step=252 epoch=1 metrics={'time_sample_batch': 0.00011102358500162761, 'time_algorithm_update': 0.004638626461937314, 'loss': 1.6706019881225767, 'time_step': 0.0048246487738594175, 'environment': 2.3999999999999995, 'advantage': -2.0480377657593314, 'td_error': 1.0484360598948828, 'value_scale': 0.5476946946530835} step=252
2023-03-14 18:45:17 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_252.pt


Epoch 2/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:20 [info     ] DiscreteCQL_20230314184515: epoch=2 step=504 epoch=2 metrics={'time_sample_batch': 0.00010811809509519547, 'time_algorithm_update': 0.00476988156636556, 'loss': 1.5411600409046051, 'time_step': 0.0049451977487594365, 'environment': 2.3999999999999995, 'advantage': -1.9709590790789915, 'td_error': 1.2673063936148248, 'value_scale': 0.4661461439724073} step=504
2023-03-14 18:45:20 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_504.pt


Epoch 3/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:22 [info     ] DiscreteCQL_20230314184515: epoch=3 step=756 epoch=3 metrics={'time_sample_batch': 0.00010149535678681873, 'time_algorithm_update': 0.004306877416277688, 'loss': 1.5132973733402433, 'time_step': 0.004469371977306548, 'environment': 2.3999999999999995, 'advantage': -1.9335201326630655, 'td_error': 1.4207385488825641, 'value_scale': 0.48433291861553274} step=756
2023-03-14 18:45:22 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_756.pt


Epoch 4/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:25 [info     ] DiscreteCQL_20230314184515: epoch=4 step=1008 epoch=4 metrics={'time_sample_batch': 9.622744151524134e-05, 'time_algorithm_update': 0.004342947687421527, 'loss': 1.501393989911155, 'time_step': 0.004492360448080396, 'environment': 2.3999999999999995, 'advantage': -1.9226060414473347, 'td_error': 1.4679456174995222, 'value_scale': 0.4349085952023634} step=1008
2023-03-14 18:45:25 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_1008.pt


Epoch 5/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:27 [info     ] DiscreteCQL_20230314184515: epoch=5 step=1260 epoch=5 metrics={'time_sample_batch': 9.588022080678789e-05, 'time_algorithm_update': 0.004232202257428851, 'loss': 1.4958202895664034, 'time_step': 0.004381706790318565, 'environment': 2.3999999999999995, 'advantage': -1.8358247695543153, 'td_error': 1.440836337100968, 'value_scale': 0.43227290682762814} step=1260
2023-03-14 18:45:27 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_1260.pt


Epoch 6/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:29 [info     ] DiscreteCQL_20230314184515: epoch=6 step=1512 epoch=6 metrics={'time_sample_batch': 9.464649927048456e-05, 'time_algorithm_update': 0.004205715088617234, 'loss': 1.4896837051898715, 'time_step': 0.0043542224263388015, 'environment': 2.3999999999999995, 'advantage': -1.917660084491656, 'td_error': 1.5131757367771783, 'value_scale': 0.4643874079004575} step=1512
2023-03-14 18:45:29 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_1512.pt


Epoch 7/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:32 [info     ] DiscreteCQL_20230314184515: epoch=7 step=1764 epoch=7 metrics={'time_sample_batch': 9.745264810229105e-05, 'time_algorithm_update': 0.004338896463787745, 'loss': 1.4852219725412035, 'time_step': 0.004488284625704326, 'environment': 2.3999999999999995, 'advantage': -2.045835791267471, 'td_error': 1.633260425717608, 'value_scale': 0.48384190170822133} step=1764
2023-03-14 18:45:32 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_1764.pt


Epoch 8/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:34 [info     ] DiscreteCQL_20230314184515: epoch=8 step=2016 epoch=8 metrics={'time_sample_batch': 0.00010483983963254898, 'time_algorithm_update': 0.004513647821214464, 'loss': 1.4829512900776334, 'time_step': 0.0046793146738930355, 'environment': 2.3999999999999995, 'advantage': -2.0076549885074138, 'td_error': 1.5920502779054697, 'value_scale': 0.4702640719086041} step=2016
2023-03-14 18:45:34 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_2016.pt


Epoch 9/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:37 [info     ] DiscreteCQL_20230314184515: epoch=9 step=2268 epoch=9 metrics={'time_sample_batch': 0.00010916164943150111, 'time_algorithm_update': 0.0049335672741844545, 'loss': 1.4794568362690153, 'time_step': 0.005107123700399247, 'environment': 2.3999999999999995, 'advantage': -1.9489484276878366, 'td_error': 1.5420630119413519, 'value_scale': 0.4633314121061772} step=2268
2023-03-14 18:45:37 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_2268.pt


Epoch 10/10:   0%|          | 0/252 [00:00<?, ?it/s]

2023-03-14 18:45:39 [info     ] DiscreteCQL_20230314184515: epoch=10 step=2520 epoch=10 metrics={'time_sample_batch': 9.992955222962395e-05, 'time_algorithm_update': 0.0043798259326389855, 'loss': 1.477288266023, 'time_step': 0.00453647356184702, 'environment': 2.3999999999999995, 'advantage': -2.073317572352768, 'td_error': 1.6343508941087483, 'value_scale': 0.4872829825090255} step=2520
2023-03-14 18:45:39 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314184515/model_2520.pt


[(1,
  {'time_sample_batch': 0.00011102358500162761,
   'time_algorithm_update': 0.004638626461937314,
   'loss': 1.6706019881225767,
   'time_step': 0.0048246487738594175,
   'environment': 2.3999999999999995,
   'advantage': -2.0480377657593314,
   'td_error': 1.0484360598948828,
   'value_scale': 0.5476946946530835}),
 (2,
  {'time_sample_batch': 0.00010811809509519547,
   'time_algorithm_update': 0.00476988156636556,
   'loss': 1.5411600409046051,
   'time_step': 0.0049451977487594365,
   'environment': 2.3999999999999995,
   'advantage': -1.9709590790789915,
   'td_error': 1.2673063936148248,
   'value_scale': 0.4661461439724073}),
 (3,
  {'time_sample_batch': 0.00010149535678681873,
   'time_algorithm_update': 0.004306877416277688,
   'loss': 1.5132973733402433,
   'time_step': 0.004469371977306548,
   'environment': 2.3999999999999995,
   'advantage': -1.9335201326630655,
   'td_error': 1.4207385488825641,
   'value_scale': 0.48433291861553274}),
 (4,
  {'time_sample_batch': 9.6

In [206]:
episode = dataset_trained.episodes[5]
print(episode.actions)
print(episode.observations)
print(episode.rewards)

[2 4 0 0 0 0 4 2 1 3 4]
[[0.]
 [0.]
 [2.]
 [1.]
 [0.]
 [0.]
 [0.]
 [2.]
 [2.]
 [3.]
 [3.]]
[-0.05  0.2  -0.2  -0.2  -0.05 -0.05  0.2  -0.05 -0.5  -0.05  2.  ]


In [179]:
for episode in dataset_trained.episodes:
    print(episode.actions)

[4 2 4 4]
[3 1 0 4 3 1 2 2 3 1 0 0 3 0 1 2 0 4 1 1]
[0 2 2 2 2 1 1 0 0 0 0 2 4 1 1 3 4]
[1 3 1 0 0 1 1 1 1 0 0 0 2 3 4 4]
[0 4 1 3 4]
[3 1 4 2 0 1 4]
[0 3 4 4 2 4]
[0 4 3 4 2 2 4]
[2 3 4 4 1]
[2 1 0 1 1 0 3 1 2 1 0 0 3 2 1 0 2 0 2 3 1 0 3 3 4 1 3 4]
[1 0 2 3 2 2 2 4 2 4 4]
[0 4 2 1 3 3 3 3 2 2 4]
[1 0 4 2 4 2 2 1]
[2 3 0 3 2 4 2 2 4 0 0 4 4]
[3 0 3 0 2 0 3 1 1 4 3 3 2 2 1]
[3 4 3 2 3 0 2 2 4 0 3 0 3 3 1 3 1 1 3 0 4]
[4 1 2 2 3 4]
[1 2 0 4 0 3 2 1 3 4 0 0 3 4 4]
[4 3 4 0 3 3 4]
[1 0 4 4 2 2 4]
[4 1 1 0 0 3 3 1 4]
[3 3 4 0 1 1 3 2 4]
[1 1 0 3 0 3 1 0 4 1 3 0 0 0 2 2 0 1 2 0 1 0 0 3 4 0 0 0 3 0 0 1 0 1 3 1 4
 1]
[4 1 4]
[2 4 1 4]
[3 2 3 2 3 1 4 2 1 0 4]
[2 1 1 1 3 1 4]
[4 2 0 1 2 2 2 4 1]
[1 4 4]
[4 3 0 1 3 1 1 4]
[3 4 2 1 2 1 2 1]
[0 4 3 3 4 3 0 2 4]
[2 1 1 0 0 2 0 2 2 1 2 2 2 4 2 3 2 3 2 3 0 3 0 0 0 3 0 4 1 2 1 4]
[1 2 3 3 4 1 4]
[1 0 2 0 2 2 1 1 3 4 4]
[1 2 2 2 3 4 4]
[1 2 4 1 0 1 0 3 1 0 1 4]
[1 0 3 4 1 0 0 0 0 3 3 4 0 4 0 1 3 3 0 2 3 1 2 0 4 0 3 0 2 1 1 2 0 3 2 1 1]
[4 1 3 3 4]
[4 2 

In [130]:
dataset_trained.compute_stats()

{'return': {'mean': 2.3999176,
  'std': 0.013471858,
  'min': 0.2,
  'max': 2.4,
  'histogram': (array([    1,     0,     0,     0,     0,     0,     0,     0,     0,
              0,     0,     0,     0,     0,     0,     0,     0,     0,
              0, 26666]),
   array([0.2       , 0.31      , 0.42000002, 0.53000003, 0.64000005,
          0.75      , 0.86      , 0.97      , 1.08      , 1.19      ,
          1.3000001 , 1.4100001 , 1.5200001 , 1.6300001 , 1.74      ,
          1.85      , 1.96      , 2.0700002 , 2.18      , 2.2900002 ,
          2.4       ], dtype=float32))},
 'reward': {'mean': 0.7999924,
  'std': 0.8485256,
  'min': 0.2,
  'max': 2.0,
  'histogram': (array([53333,     0,     0,     0,     0,     0,     0,     0,     0,
              0,     0,     0,     0,     0,     0,     0,     0,     0,
              0, 26666]),
   array([0.2 , 0.29, 0.38, 0.47, 0.56, 0.65, 0.74, 0.83, 0.92, 1.01, 1.1 ,
          1.19, 1.28, 1.37, 1.46, 1.55, 1.64, 1.73, 1.82, 1.91, 2.  ],
  

In [207]:
evaluate_scorer = evaluate_on_environment(env, n_trials=1)
mean_episode_return = evaluate_scorer(cql)
mean_episode_return

2.4

In [208]:
for i in range(5):
    obs = np.array([i])
    action = cql.predict([obs])[0]
    print(action)

4
1
4
4
4


### Custom environment

In [1]:
import gym
from gym.spaces import Discrete, Box
import numpy as np
import os
import random

class SimpleCorridor_d3rlpy(gym.Env):
    """Example of a custom env in which you have to walk down a corridor.
    Get a reward of -0.1 if you are not at the end, a random reward that is positive if you do.
    Move +1 if you move forward, -1 if you move backward. The total length is 5.
    We should want to reach the end in 5 steps in the perfectly trained world. 
    You can configure the length of the corridor via the env config."""

    def __init__(self, config):
        self.end_pos = config["corridor_length"]
        self.cur_pos = 0
        self.action_space = Discrete(5)
        self.observation_space = Box(0.0, self.end_pos, shape=(1,), dtype=np.float32)
        # Set the seed. This is only used for the final (reach goal) reward.
        self.reset()

    def reset(self, *, seed=None, options=None):
        random.seed(seed)
        self.cur_pos = 0
        return np.array([self.cur_pos])

    def step(self, action):
        assert action in [0, 1, 2, 3, 4], action
        # backward step
        if action == 0 and self.cur_pos > 0:
            self.cur_pos -= 1
            reward = -0.2
        # forward step
        elif action == 1:
            self.cur_pos += 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = 0.1
        # double speed
        elif action == 4:
            if(self.cur_pos <= 3):
                self.cur_pos += 2
            else:
                self.cur_pos += 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = 0.2
        # left or right
        else:
            self.cur_pos = self.cur_pos
            reward = -0.05
            
        done = truncated = self.cur_pos >= self.end_pos
        if(done):
            reward = 2
        # Produce a random reward when we reach the goal.
        return (
            np.array([self.cur_pos]),
            reward, # Setting to 2 instead of random reward has no real impact
            done,
            {},
        )
    
    def check_if_water(self):
        # This is water and will get a negative reward
        if(self.cur_pos == 3):
            return True
        else:
            return False

config={"corridor_length": 5}
env = SimpleCorridor_d3rlpy(config=config)

In [220]:
class SimpleCorridor_stochastic(gym.Env):
    """Example of a custom env in which you have to walk down a corridor.
    Get a reward of -0.1 if you are not at the end, a random reward that is positive if you do.
    Move +1 if you move forward, -1 if you move backward. The total length is 5.
    We should want to reach the end in 5 steps in the perfectly trained world. 
    You can configure the length of the corridor via the env config."""

    def __init__(self, config):
        self.end_pos = config["corridor_length"]
        self.cur_pos = 0
        self.action_space = Discrete(5)
        self.observation_space = Box(0.0, self.end_pos, shape=(1,), dtype=np.float32)
        # Set the seed. This is only used for the final (reach goal) reward.
        self.reset()

    def reset(self, *, seed=None, options=None):
        random.seed(seed)
        self.cur_pos = 0
        return np.array([self.cur_pos])

    def step(self, action):
        assert action in [0, 1, 2, 3, 4], action
        
        # backward step
        if action == 0 and self.cur_pos > 0:
            self.cur_pos -= 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = -0.2
            #reward = -0.2
        # forward step
        elif action == 1:
            self.cur_pos += 1
            if(self.check_if_water()):
                reward = -0.5
            else:
                reward = 0.1
        # double speed
        elif action == 4:
            inc = np.random.choice([1, 2], 1, p=[0.1, 0.9])[0]
            if(self.cur_pos <= 3):
                self.cur_pos += inc
                reward = 0.2 if inc == 2 else 0.1
            else:
                self.cur_pos += 1
                reward = 0.2
                
            if(self.check_if_water()):
                reward = -0.5  
        # left or right
        else:
            self.cur_pos = self.cur_pos
            reward = -0.05
            
        done = truncated = self.cur_pos >= self.end_pos
        
        if(done):
            reward = 2
        # Produce a random reward when we reach the goal.
        return (
            np.array([self.cur_pos]),
            reward, # Setting to 2 instead of random reward has no real impact
            truncated,
            {},
        )
    
    def check_if_water(self):
        # This is water and will get a negative reward
        if(self.cur_pos == 3):
            return True
        else:
            return False

In [221]:
from gym.wrappers import TimeLimit
from d3rlpy.algos import DQN
from d3rlpy.online.buffers import ReplayBuffer
from d3rlpy.online.explorers import LinearDecayEpsilonGreedy, ConstantEpsilonGreedy

config={"corridor_length": 5}
env_corridor = SimpleCorridor_stochastic(config=config)

env = TimeLimit(env_corridor, max_episode_steps=2000)

#env = gym.make('CartPole-v0')
#eval_env = gym.make('CartPole-v0')

# setup algorithm
dqn = DQN(batch_size=128,
          learning_rate=2.5e-5,
          target_update_interval=100,
          use_gpu=False)

# setup replay buffer
buffer = ReplayBuffer(maxlen=100000, env=env)

# setup explorers
explorer = LinearDecayEpsilonGreedy(start_epsilon=1.0,
                                    end_epsilon=0.1,
                                    duration=10000)

# start training
dqn.fit_online(env,
               buffer,
               explorer=explorer, # you don't need this with probablistic policy algorithms
               eval_env=env,
               n_steps=6000, # the number of total steps to train.
               n_steps_per_epoch=1000,
               update_interval=10) # update parameters every 10 steps.

# export online dataset as MDPDataset
dataset_online = buffer.to_mdp_dataset()

# save MDPDataset
dataset_online.dump("trained_corridor_stochastic_policy_dataset_6000_online.h5")

2023-03-14 20:51:18 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230314205118
2023-03-14 20:51:18 [debug    ] Building model...
2023-03-14 20:51:18 [debug    ] Model has been built.
2023-03-14 20:51:18 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230314205118/params.json params={'action_scaler': None, 'batch_size': 128, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 2.5e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/6000 [00:00<?, ?it/s]

2023-03-14 20:51:19 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314205118/model_1000.pt
2023-03-14 20:51:19 [info     ] DQN_online_20230314205118: epoch=1 step=1000 epoch=1 metrics={'time_inference': 0.0004667389392852783, 'time_environment_step': 2.6211023330688476e-05, 'time_step': 0.0009442102909088134, 'rollout_return': 1.3494444444444447, 'time_sample_batch': 0.00017553296956149015, 'time_algorithm_update': 0.00444207408211448, 'loss': 0.09772586449980736, 'evaluation': 2.3199999999999994} step=1000
2023-03-14 20:51:20 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230314205118/model_2000.pt
2023-03-14 20:51:21 [info     ] DQN_online_20230314205118: epoch=2 step=2000 epoch=2 metrics={'time_inference': 0.0005018572807312012, 'time_environment_step': 3.40578556060791e-05, 'rollout_return': 1.6169421487603302, 'time_step': 0.0010680263042449952, 'time_sample_batch': 0.00019301652908325197, 'time_algorithm_update': 0.004621822834014893, 'l

In [222]:
for episode in dataset_online.episodes:
    print(episode.actions)

[2 3 0 4 1 4]
[2 1 4 4]
[2 2 0 2 2 3 3 3 0 3 1 2 4 0 2 0 4 1 0 0 0 4 3 1 4]
[2 4 0 4 4 1]
[4 4 3 3 1]
[0 4 4 2 0 0 0 2 0 2 2 4 3 0 1 0 0 2 2 0 2 3 2 2 0 2 0 1 3 0 3 2 1 1 3 4 2
 1 0 0 0 1 4 2 0 2 2 3 1 1]
[1 4 4 3 4]
[4 1 0 4 4]
[4 0 1 2 0 0 4 1 0 2 2 1 4]
[4 0 4 4]
[3 1 0 2 0 1 3 1 0 2 1 3 1 1 1]
[2 3 1 1 2 1 0 4 0 0 1 1 3 0 3 1 0 1 3 4]
[4 1 2 1 3 1]
[0 4 3 0 0 1 0 1 1 3 2 3 0 1 4 3 3 0 0 0 1 4 2 3 2 2 2 2 3 3 4]
[1 4 3 4 4]
[2 2 3 3 0 1 0 0 0 3 1 0 3 0 4 1 2 3 3 2 1 2 3 4]
[4 0 4 2 2 4]
[3 2 4 2 0 3 4 4]
[2 0 3 0 3 3 0 1 0 3 3 3 2 0 3 0 2 0 0 4 2 2 3 4 0 2 2 4]
[1 3 3 3 3 0 1 3 3 1 2 3 4 0 0 1 0 0 0 4 2 3 4 0 1 3 4]
[0 3 2 4 0 3 1 1 4]
[3 3 3 0 2 4 3 4 3 2 4]
[4 0 1 2 2 3 3 3 3 4 3 4]
[3 0 0 3 2 2 1 4 1 2 4]
[0 3 3 4 1 0 0 4 0 4 4]
[3 0 0 1 1 2 4 2 1]
[4 0 0 2 3 3 4 3 4 1]
[0 1 1 4 1]
[4 1 2 0 3 0 3 0 0 1 4 0 3 0 0 3 3 3 3 0 1 2 2 0 0 0 3 4 0 3 4 4]
[2 4 2 2 4 1]
[2 4 3 3 2 3 0 2 2 2 2 2 1 3 4 1]
[1 2 0 1 4 0 4 3 1]
[1 3 2 1 2 3 2 4 2 0 0 4 2 0 0 4 2 2 4]
[1 3 2 3 4 1 1 0 2 4]
[1 1 

In [263]:
evaluate_scorer = evaluate_on_environment(env, n_trials=1)
mean_episode_return = evaluate_scorer(cql_stochastic)
mean_episode_return

2.4

In [262]:
for i in range(5):
    obs = np.array([i])
    action = cql_stochastic.predict([obs])[0]
    print(action)

4
1
4
4
4


In [225]:
dataset_trained = dataset.load("trained_corridor_stochastic_policy_dataset_6000_online.h5")
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.3)

In [254]:
from d3rlpy.algos import DQN
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer

# if you don't use GPU, set use_gpu=False instead.
dqn1 = DQN(batch_size=128,
          learning_rate=2.5e-5,
          target_update_interval=100,
          use_gpu=False)

# initialize neural networks with the given observation shape and action size.
# this is not necessary when you directly call fit or fit_online method.
dqn1.build_with_dataset(dataset_trained)

# calculate metrics with test dataset
td_error = td_error_scorer(dqn1, test_episodes)

# set environment in scorer function, returns mean return per episode
evaluate_scorer = evaluate_on_environment(env)

# evaluate algorithm on the environment
rewards = evaluate_scorer(dqn1)

# Use the output of learning, this is the mean reward and the most useful metric to measure learning
dqn1.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'td_error': td_error_scorer,
            'advantage': discounted_sum_of_advantage_scorer,
            'value_scale': average_value_estimation_scorer,
            'environment': evaluate_scorer
        })

2023-03-14 21:16:01 [debug    ] RoundIterator is selected.
2023-03-14 21:16:01 [info     ] Directory is created at d3rlpy_logs/DQN_20230314211601
2023-03-14 21:16:01 [warning  ] Skip building models since they're already built.
2023-03-14 21:16:01 [info     ] Parameters are saved to d3rlpy_logs/DQN_20230314211601/params.json params={'action_scaler': None, 'batch_size': 128, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 2.5e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 100, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


Epoch 1/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:02 [info     ] DQN_20230314211601: epoch=1 step=33 epoch=1 metrics={'time_sample_batch': 0.00034640052101828837, 'time_algorithm_update': 0.005879618904807351, 'loss': 0.2890530824661255, 'time_step': 0.006334167538267194, 'td_error': 0.5047065423580975, 'advantage': -0.8995868796343676, 'value_scale': 0.29699211571883233, 'environment': 1.8} step=33
2023-03-14 21:16:02 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_33.pt


Epoch 2/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:04 [info     ] DQN_20230314211601: epoch=2 step=66 epoch=2 metrics={'time_sample_batch': 0.000310254819465406, 'time_algorithm_update': 0.005703528722127278, 'loss': 0.15543649300481333, 'time_step': 0.006111051097060695, 'td_error': 0.3651020040824381, 'advantage': -1.2757280428786089, 'value_scale': 0.5251958018554913, 'environment': 2.3199999999999994} step=66
2023-03-14 21:16:04 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_66.pt


Epoch 3/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:05 [info     ] DQN_20230314211601: epoch=3 step=99 epoch=3 metrics={'time_sample_batch': 0.00028665860493977863, 'time_algorithm_update': 0.0054805062033913355, 'loss': 0.07384830239144238, 'time_step': 0.005897637569543087, 'td_error': 0.4619677232645753, 'advantage': -1.959397080674803, 'value_scale': 0.7927307365525827, 'environment': 2.3999999999999995} step=99
2023-03-14 21:16:05 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_99.pt


Epoch 4/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:06 [info     ] DQN_20230314211601: epoch=4 step=132 epoch=4 metrics={'time_sample_batch': 0.0003422101338704427, 'time_algorithm_update': 0.005716049309932824, 'loss': 0.16508958138751262, 'time_step': 0.00616665319962935, 'td_error': 0.5044802544531579, 'advantage': -2.122430742259591, 'value_scale': 1.045183839331591, 'environment': 2.2699999999999996} step=132
2023-03-14 21:16:06 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_132.pt


Epoch 5/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:08 [info     ] DQN_20230314211601: epoch=5 step=165 epoch=5 metrics={'time_sample_batch': 0.000337680180867513, 'time_algorithm_update': 0.006190726251313181, 'loss': 0.08226763919898958, 'time_step': 0.006643793799660422, 'td_error': 0.5103397963650592, 'advantage': -2.092527476393524, 'value_scale': 1.2272837168806505, 'environment': 2.2499999999999996} step=165
2023-03-14 21:16:08 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_165.pt


Epoch 6/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:09 [info     ] DQN_20230314211601: epoch=6 step=198 epoch=6 metrics={'time_sample_batch': 0.0003568548144716205, 'time_algorithm_update': 0.006043426918260979, 'loss': 0.045745589403492035, 'time_step': 0.006515871394764294, 'td_error': 0.4859936051961278, 'advantage': -1.988018763395136, 'value_scale': 1.3244418708908803, 'environment': 2.1799999999999997} step=198
2023-03-14 21:16:09 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_198.pt


Epoch 7/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:10 [info     ] DQN_20230314211601: epoch=7 step=231 epoch=7 metrics={'time_sample_batch': 0.0003027915954589844, 'time_algorithm_update': 0.005785616961392489, 'loss': 0.1796925842310443, 'time_step': 0.006184866934111624, 'td_error': 0.489564662766025, 'advantage': -1.795355629869317, 'value_scale': 1.4997455003864402, 'environment': 2.1} step=231
2023-03-14 21:16:10 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_231.pt


Epoch 8/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:12 [info     ] DQN_20230314211601: epoch=8 step=264 epoch=8 metrics={'time_sample_batch': 0.0003298990654222893, 'time_algorithm_update': 0.005736105369799065, 'loss': 0.1284815242344683, 'time_step': 0.006172620888912316, 'td_error': 0.47288097700263626, 'advantage': -1.599040531134627, 'value_scale': 1.6248546538700548, 'environment': 2.3199999999999994} step=264
2023-03-14 21:16:12 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_264.pt


Epoch 9/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:13 [info     ] DQN_20230314211601: epoch=9 step=297 epoch=9 metrics={'time_sample_batch': 0.00036170988371877957, 'time_algorithm_update': 0.0064723274924538355, 'loss': 0.10140044061523495, 'time_step': 0.006969618074821703, 'td_error': 0.41214882939418696, 'advantage': -1.4320772769783927, 'value_scale': 1.6567416505417139, 'environment': 2.2399999999999998} step=297
2023-03-14 21:16:13 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_297.pt


Epoch 10/10:   0%|          | 0/33 [00:00<?, ?it/s]

2023-03-14 21:16:14 [info     ] DQN_20230314211601: epoch=10 step=330 epoch=10 metrics={'time_sample_batch': 0.0003482934200402462, 'time_algorithm_update': 0.005911436947909268, 'loss': 0.17027455387693463, 'time_step': 0.006377892060713334, 'td_error': 0.4136211067090206, 'advantage': -1.3125564194076937, 'value_scale': 1.7720881764329375, 'environment': 2.3999999999999995} step=330
2023-03-14 21:16:14 [info     ] Model parameters are saved to d3rlpy_logs/DQN_20230314211601/model_330.pt


[(1,
  {'time_sample_batch': 0.00034640052101828837,
   'time_algorithm_update': 0.005879618904807351,
   'loss': 0.2890530824661255,
   'time_step': 0.006334167538267194,
   'td_error': 0.5047065423580975,
   'advantage': -0.8995868796343676,
   'value_scale': 0.29699211571883233,
   'environment': 1.8}),
 (2,
  {'time_sample_batch': 0.000310254819465406,
   'time_algorithm_update': 0.005703528722127278,
   'loss': 0.15543649300481333,
   'time_step': 0.006111051097060695,
   'td_error': 0.3651020040824381,
   'advantage': -1.2757280428786089,
   'value_scale': 0.5251958018554913,
   'environment': 2.3199999999999994}),
 (3,
  {'time_sample_batch': 0.00028665860493977863,
   'time_algorithm_update': 0.0054805062033913355,
   'loss': 0.07384830239144238,
   'time_step': 0.005897637569543087,
   'td_error': 0.4619677232645753,
   'advantage': -1.959397080674803,
   'value_scale': 0.7927307365525827,
   'environment': 2.3999999999999995}),
 (4,
  {'time_sample_batch': 0.00034221013387044

In [261]:
from d3rlpy.dataset import MDPDataset
from d3rlpy.algos import DiscreteCQL
from d3rlpy.datasets import get_cartpole
from sklearn.model_selection import train_test_split
from d3rlpy.metrics.scorer import discounted_sum_of_advantage_scorer
from d3rlpy.metrics.scorer import evaluate_on_environment
from d3rlpy.metrics.scorer import td_error_scorer
from d3rlpy.metrics.scorer import average_value_estimation_scorer

#dataset, env = get_cartpole()

config={"corridor_length": 5}
env_corridor = SimpleCorridor_stochastic(config=config)

env = TimeLimit(env_corridor, max_episode_steps=2000)

# setup CQL algorithm (discrete version)
cql_stochastic = DiscreteCQL(n_frames=4, use_gpu=False)

# split train and test episodes
#dataset_trained = dataset.load("trained_policy_dataset_80000.h5")
dataset_trained = dataset.load("trained_corridor_stochastic_policy_dataset_6000_online.h5")
train_episodes, test_episodes = train_test_split(dataset_trained, test_size=0.2)

# start training
cql_stochastic.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=10,
        scorers={
            'environment': evaluate_on_environment(env), # Cartpole environment
            'advantage': discounted_sum_of_advantage_scorer, # smaller is better
            'td_error': td_error_scorer, # smaller is better
            'value_scale': average_value_estimation_scorer # smaller is better
        })

2023-03-14 21:20:42 [debug    ] RoundIterator is selected.
2023-03-14 21:20:42 [info     ] Directory is created at d3rlpy_logs/DiscreteCQL_20230314212042
2023-03-14 21:20:42 [debug    ] Building models...
2023-03-14 21:20:42 [debug    ] Models have been built.
2023-03-14 21:20:42 [info     ] Parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/params.json params={'action_scaler': None, 'alpha': 1.0, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 4, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DiscreteCQL', 'observation

Epoch 1/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:43 [info     ] DiscreteCQL_20230314212042: epoch=1 step=149 epoch=1 metrics={'time_sample_batch': 0.00012951249244229105, 'time_algorithm_update': 0.005374508416092636, 'loss': 1.6036005948213923, 'time_step': 0.005581401338513266, 'environment': 2.34, 'advantage': -2.8820603118363266, 'td_error': 0.5922182975316892, 'value_scale': 0.8685661791975982} step=149
2023-03-14 21:20:43 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_149.pt


Epoch 2/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:45 [info     ] DiscreteCQL_20230314212042: epoch=2 step=298 epoch=2 metrics={'time_sample_batch': 0.00010570903752474176, 'time_algorithm_update': 0.00450339573341728, 'loss': 1.5781406572201109, 'time_step': 0.004667882151251672, 'environment': 2.25, 'advantage': -2.5467079932365215, 'td_error': 0.5226088940134123, 'value_scale': 0.7957506918537072} step=298
2023-03-14 21:20:45 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_298.pt


Epoch 3/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:46 [info     ] DiscreteCQL_20230314212042: epoch=3 step=447 epoch=3 metrics={'time_sample_batch': 0.00012388965427475486, 'time_algorithm_update': 0.004990585698377366, 'loss': 1.574138202123194, 'time_step': 0.005182035817395921, 'environment': 2.3999999999999995, 'advantage': -2.8166430487101364, 'td_error': 0.6135596439418888, 'value_scale': 0.8652555497935395} step=447
2023-03-14 21:20:46 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_447.pt


Epoch 4/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:48 [info     ] DiscreteCQL_20230314212042: epoch=4 step=596 epoch=4 metrics={'time_sample_batch': 0.00012174388706284081, 'time_algorithm_update': 0.0054850450298130115, 'loss': 1.5707992627316674, 'time_step': 0.005683193270792097, 'environment': 2.32, 'advantage': -2.851890783583336, 'td_error': 0.6401235299518566, 'value_scale': 0.8390436306401016} step=596
2023-03-14 21:20:48 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_596.pt


Epoch 5/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:49 [info     ] DiscreteCQL_20230314212042: epoch=5 step=745 epoch=5 metrics={'time_sample_batch': 0.00011330162918807677, 'time_algorithm_update': 0.004996150932055992, 'loss': 1.5671061245387032, 'time_step': 0.005174696045433915, 'environment': 2.2399999999999998, 'advantage': -2.893766367196263, 'td_error': 0.6652049774875789, 'value_scale': 0.8711542911291902} step=745
2023-03-14 21:20:49 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_745.pt


Epoch 6/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:50 [info     ] DiscreteCQL_20230314212042: epoch=6 step=894 epoch=6 metrics={'time_sample_batch': 0.00010025581257455301, 'time_algorithm_update': 0.004498086519689368, 'loss': 1.5651575446928907, 'time_step': 0.004654762728902318, 'environment': 2.3299999999999996, 'advantage': -3.001637504884407, 'td_error': 0.7160234809960135, 'value_scale': 0.8913642590427223} step=894
2023-03-14 21:20:50 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_894.pt


Epoch 7/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:52 [info     ] DiscreteCQL_20230314212042: epoch=7 step=1043 epoch=7 metrics={'time_sample_batch': 0.00010804041920092282, 'time_algorithm_update': 0.004740815834711062, 'loss': 1.5625344978882962, 'time_step': 0.004903588519000367, 'environment': 2.3299999999999996, 'advantage': -2.6589235944745804, 'td_error': 0.5929305300883511, 'value_scale': 0.8272351685278361} step=1043
2023-03-14 21:20:52 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_1043.pt


Epoch 8/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:53 [info     ] DiscreteCQL_20230314212042: epoch=8 step=1192 epoch=8 metrics={'time_sample_batch': 0.00011225354751484506, 'time_algorithm_update': 0.004956567047426364, 'loss': 1.5617464916818093, 'time_step': 0.005131505479748617, 'environment': 2.27, 'advantage': -2.7038472584569453, 'td_error': 0.6164781498356842, 'value_scale': 0.8343525600978752} step=1192
2023-03-14 21:20:53 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_1192.pt


Epoch 9/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:55 [info     ] DiscreteCQL_20230314212042: epoch=9 step=1341 epoch=9 metrics={'time_sample_batch': 0.00010655710361147887, 'time_algorithm_update': 0.004639155112656971, 'loss': 1.5600991065070133, 'time_step': 0.00480502243810052, 'environment': 2.3999999999999995, 'advantage': -2.6733066873445877, 'td_error': 0.6127600681323564, 'value_scale': 0.819173668760782} step=1341
2023-03-14 21:20:55 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_1341.pt


Epoch 10/10:   0%|          | 0/149 [00:00<?, ?it/s]

2023-03-14 21:20:56 [info     ] DiscreteCQL_20230314212042: epoch=10 step=1490 epoch=10 metrics={'time_sample_batch': 0.00011280559053357015, 'time_algorithm_update': 0.004914429363788374, 'loss': 1.5579191966344845, 'time_step': 0.005090590291375281, 'environment': 2.3999999999999995, 'advantage': -2.9194020809580263, 'td_error': 0.6798155551852949, 'value_scale': 0.8801705754830752} step=1490
2023-03-14 21:20:56 [info     ] Model parameters are saved to d3rlpy_logs/DiscreteCQL_20230314212042/model_1490.pt


[(1,
  {'time_sample_batch': 0.00012951249244229105,
   'time_algorithm_update': 0.005374508416092636,
   'loss': 1.6036005948213923,
   'time_step': 0.005581401338513266,
   'environment': 2.34,
   'advantage': -2.8820603118363266,
   'td_error': 0.5922182975316892,
   'value_scale': 0.8685661791975982}),
 (2,
  {'time_sample_batch': 0.00010570903752474176,
   'time_algorithm_update': 0.00450339573341728,
   'loss': 1.5781406572201109,
   'time_step': 0.004667882151251672,
   'environment': 2.25,
   'advantage': -2.5467079932365215,
   'td_error': 0.5226088940134123,
   'value_scale': 0.7957506918537072}),
 (3,
  {'time_sample_batch': 0.00012388965427475486,
   'time_algorithm_update': 0.004990585698377366,
   'loss': 1.574138202123194,
   'time_step': 0.005182035817395921,
   'environment': 2.3999999999999995,
   'advantage': -2.8166430487101364,
   'td_error': 0.6135596439418888,
   'value_scale': 0.8652555497935395}),
 (4,
  {'time_sample_batch': 0.00012174388706284081,
   'time_al

In [15]:
env

In [16]:
env.reset()

array([0])

In [5]:
action = env.action_space.sample()
obs, reward, done, info = env.step(action)

In [6]:
reward

-0.05

In [19]:
obs.astype

<function ndarray.astype>

In [41]:
import d3rlpy

# setup algorithm
random_policy = d3rlpy.algos.DiscreteRandomPolicy()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)

# start data collection
random_policy.collect(corridor_env, buffer, n_steps=100000)

# export as MDPDataset
dataset = buffer.to_mdp_dataset()

# save MDPDataset
dataset.dump("random_policy_corridor_dataset.h5")

2023-03-13 21:33:12 [debug    ] Building model...
2023-03-13 21:33:12 [debug    ] Model has been built.


  0%|          | 0/100000 [00:00<?, ?it/s]

### Fit online with custom environment

In [113]:
# setup algorithm
dqn = d3rlpy.algos.DQN()

# prepare experience replay buffer
buffer = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)

# prepare exploration strategy if necessary
explorer = d3rlpy.online.explorers.ConstantEpsilonGreedy(0.3)

# start data collection
dqn.fit_online(env, buffer, n_steps=100000)

2023-03-10 14:47:32 [info     ] Directory is created at d3rlpy_logs/DQN_online_20230310144732
2023-03-10 14:47:32 [debug    ] Building model...
2023-03-10 14:47:32 [debug    ] Model has been built.
2023-03-10 14:47:32 [info     ] Parameters are saved to d3rlpy_logs/DQN_online_20230310144732/params.json params={'action_scaler': None, 'batch_size': 32, 'encoder_factory': {'type': 'default', 'params': {'activation': 'relu', 'use_batch_norm': False, 'dropout_rate': None}}, 'gamma': 0.99, 'generated_maxlen': 100000, 'learning_rate': 6.25e-05, 'n_critics': 1, 'n_frames': 1, 'n_steps': 1, 'optim_factory': {'optim_cls': 'Adam', 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False}, 'q_func_factory': {'type': 'mean', 'params': {'share_encoder': False}}, 'real_ratio': 1.0, 'reward_scaler': None, 'scaler': None, 'target_update_interval': 8000, 'use_gpu': None, 'algorithm': 'DQN', 'observation_shape': (1,), 'action_size': 5}


  0%|          | 0/100000 [00:00<?, ?it/s]

2023-03-10 14:48:12 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230310144732/model_10000.pt
2023-03-10 14:48:12 [info     ] DQN_online_20230310144732: epoch=1 step=10000 epoch=1 metrics={'time_inference': 0.0005066035032272339, 'time_environment_step': 1.2087106704711915e-05, 'time_step': 0.003947368121147155, 'time_sample_batch': 0.0001312981424975613, 'time_algorithm_update': 0.00323518305445813, 'loss': 1.49070639037667e-06} step=10000
2023-03-10 14:48:50 [info     ] Model parameters are saved to d3rlpy_logs/DQN_online_20230310144732/model_20000.pt
2023-03-10 14:48:50 [info     ] DQN_online_20230310144732: epoch=2 step=20000 epoch=2 metrics={'time_inference': 0.0004898247003555298, 'time_environment_step': 1.1702489852905273e-05, 'time_sample_batch': 0.00012943384647369385, 'time_algorithm_update': 0.003080199909210205, 'loss': 7.911302969334136e-07, 'time_step': 0.0037821149587631228} step=20000
2023-03-10 14:49:30 [info     ] Model parameters are saved to d3

In [21]:
for i in range(5):
    obs = np.array([i])
    action = dqn.predict([obs])[0]
    print(action)

4
1
4
4
4


### Fit on offline data generated from custom environment

In [147]:
#initialize the algorithm
dqn = d3rlpy.algos.DQN()

dataset2 = d3rlpy.online.buffers.ReplayBuffer(maxlen=100000, env=env)
dataset2 = buffer.to_mdp_dataset()
dataset2.load("random_policy_corridor_dataset.h5")

train_episodes, test_episodes = train_test_split(dataset2, test_size=0.2)

In [ ]:
# start training
dqn.fit(train_episodes,
        eval_episodes=test_episodes,
        n_epochs=5,
        scorers={
            'environment': evaluate_on_environment(env), # Cartpole environment
            'advantage': discounted_sum_of_advantage_scorer, # smaller is better
            'td_error': td_error_scorer, # smaller is better
            'value_scale': average_value_estimation_scorer # smaller is better
        })

In [148]:
dqn.fit(
    train_episodes,
    n_epochs=10,
    #scorers=metrics,
    eval_episodes=test_episodes,
)